airbyte-cdk/python/airbyte_cdk/sources/concurrent_source/thread_pool_manager.py

```diff
@@ -0,0 +1,97 @@
+#
+# Copyright (c) 2023 Airbyte, Inc., all rights reserved.
+#
+import logging
+import time
+from concurrent.futures import Future, ThreadPoolExecutor
+from typing import Any, Callable, List
+
+
+class ThreadPoolManager:
+    """
+    Wrapper to abstract away the threadpool and the logic to wait for pending tasks to be completed.
+    """
+
+    DEFAULT_SLEEP_TIME = 0.1
+    DEFAULT_MAX_QUEUE_SIZE = 10_000
+
+    def __init__(
+        self,
+        threadpool: ThreadPoolExecutor,
+        logger: logging.Logger,
+        max_concurrent_tasks: int = DEFAULT_MAX_QUEUE_SIZE,
+        sleep_time: float = DEFAULT_SLEEP_TIME,
+    ):
+        """
+        :param threadpool: The threadpool to use
+        :param logger: The logger to use
+        :param max_concurrent_tasks: The maximum number of tasks that can be pending at the same time
+        :param sleep_time: How long to sleep if there are too many pending tasks
+        """
+        self._threadpool = threadpool
+        self._logger = logger
+        self._max_concurrent_tasks = max_concurrent_tasks
+        self._sleep_time = sleep_time
+        self._futures: List[Future[Any]] = []
+
+    def submit(self, function: Callable[..., Any], *args: Any) -> None:
+        # Submit a task to the threadpool, waiting if there are too many pending tasks
+        self._wait_while_too_many_pending_futures(self._futures)
+        self._futures.append(self._threadpool.submit(function, *args))
+
+    def _wait_while_too_many_pending_futures(self, futures: List[Future[Any]]) -> None:
+        # Wait until the number of pending tasks is < self._max_concurrent_tasks
+        while True:
+            self._prune_futures(futures)
+            if len(futures) < self._max_concurrent_tasks:
+                break
+            self._logger.info("Main thread is sleeping because the task queue is full...")
+            time.sleep(self._sleep_time)
+
+    def _prune_futures(self, futures: List[Future[Any]]) -> None:
+        """
+        Take a list in input and remove the futures that are completed. If a future has an exception, it'll raise and kill the stream
+        operation.
+
+        Pruning this list safely relies on the assumptions that only the main thread can modify the list of futures.
+        """
+        if len(futures) < self._max_concurrent_tasks:
+            return
+
+        for index in reversed(range(len(futures))):
+            future = futures[index]
+            optional_exception = future.exception()
+            if optional_exception:
+                exception = RuntimeError(f"Failed reading with error: {optional_exception}")
+                self._stop_and_raise_exception(exception)
+
+            if future.done():
+                futures.pop(index)
```

Comment:
I think we can use [`concurrent.futures.as_completed`](https://docs.python.org/3/library/concurrent.futures.html#concurrent.futures.as_completed) to simplify this.

```suggestion
    def _wait_while_too_many_pending_futures(self, futures: List[Future[Any]]) -> None:
        # Wait until the number of pending tasks is < self._max_concurrent_tasks
        while self._too_many_pending_futures(futures):
            for future in as_completed(futures):
                try:
                    future.result()
                except Exception as e:
                    self._stop_and_raise_exception(e)
                if not self._too_many_pending_futures(futures):
                    break
            if self._too_many_pending_futures(futures):
                self._logger.info("Main thread is sleeping because the task queue is full...")
                time.sleep(self._sleep_time)

    def _too_many_pending_futures(self, futures: List[Future[Any]]) -> bool:
        return len(futures) > self._max_concurrent_tasks
```

Reply:
@clnoll I think I'm missing something from your suggestion.

We still need to remove the complete futures from list of futures because Maxime's testing showed we spent a lot of time iterating over them. That's not super surprising in hindsight because substreams have tons of partitions.

Is there a reason to think `as_completed` can iterate over them faster than we can? As far as I can tell, it's iterating over them the same way we do
```
        finished = set(
                f for f in fs
                if f._state in [CANCELLED_AND_NOTIFIED, FINISHED])
```

Reply:
Ah true, in that case I suppose we still need to be able to prune them.

Reply:
right. I'm somewhat partial to the current implementation since it seems it seems simpler than iterating over `as_completed(futures)` before pruning the futures. let me know if you disagree

Reply:
That works for me!